compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -0,0 +1,329 @@
+//! The loop invariant code motion pass moves code from inside a loop to before the loop
+//! if that code will always have the same result on every iteration of the loop.
+//!
+//! To identify a loop invariant, check whether all of an instruction's values are:
+//! - Outside of the loop
+//! - Constant
+//! - Already marked as loop invariants
+//!
+//! We also check that we are not hoisting instructions with side effects.
+use fxhash::{FxHashMap as HashMap, FxHashSet as HashSet};
+
+use crate::ssa::{
+    ir::{
+        function::{Function, RuntimeType},
+        function_inserter::FunctionInserter,
+        value::ValueId,
+    },
+    Ssa,
+};
+
+use super::{
+    constant_folding::{push_instruction, replace_result_ids},
+    unrolling::Loops,
+};
+
+impl Ssa {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn loop_invariant_code_motion(mut self) -> Ssa {
+        let brillig_functions = self
+            .functions
+            .iter_mut()
+            .filter(|(_, func)| matches!(func.runtime(), RuntimeType::Brillig(_)));
+
+        for (_, function) in brillig_functions {
+            function.loop_invariant_code_motion();
+        }
+
+        self
+    }
+}
+
+impl Function {
+    fn loop_invariant_code_motion(&mut self) {
+        Loops::find_all(self).hoist_loop_invariants(self);
+    }
+}
+
+impl Loops {
+    fn hoist_loop_invariants(self, function: &mut Function) {
```

Comment:
Can you break this up into helper functions to show the individual steps a bit better?

Reply:
I have added a new `LoopInvariantContext` struct and have broken up this method into a few smaller methods.